### PR TITLE
Enhacements.

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

- [x] Clean code.
- [x] Add new views login, sigunp, with `yiisoft/form`. 
- [x] Add tests `LoginAcceptance` and `SignupAcceptance`.
- [x] Add support `PHP 8.1` github actions. 

image

image

### CODE REVIEW

File: src/Controller/SignupController.php [hunk 1]
```diff
@@ -25,31 +26,29 @@ public function __construct(ViewRenderer $viewRenderer)
     }
 
     public function signup(
-        RequestInterface $request,
-        IdentityRepositoryInterface $identityRepository,
         EntityWriter $entityWriter,
+        IdentityRepositoryInterface $identityRepository,
+        LoginForm $loginForm,
```
Comment (BoShurik) [inline src/Controller/SignupController.php:31]: I think `LoginForm` is a data-class and should not be created by container. 
In that case probably better to move translation to the template

Comment (terabytesoftw) [reply src/Controller/SignupController.php]: done.

File: src/Controller/SignupController.php [hunk 2]
```diff
@@ -25,31 +27,31 @@ public function __construct(ViewRenderer $viewRenderer)
...
         $error = null;
 
-        if ($request->getMethod() === Method::POST) {
-            try {
-                foreach (['login', 'password'] as $name) {
-                    if (empty($body[$name])) {
-                        throw new \InvalidArgumentException(ucfirst($name) . ' is required.');
-                    }
-                }
+        $loginForm = new LoginForm($translator);
```
Comment (samdark) [inline src/Controller/SignupController.php:41]: I don't think we should reuse login form for both signup and login. Should be two separate forms.

Usually a signup form has more data: password repeat, some user data etc.

File: src/Controller/SignupController.php [hunk 3]
```diff
@@ -25,31 +27,31 @@ public function __construct(ViewRenderer $viewRenderer)
...
+            $request->getMethod() === Method::POST
+            && $loginForm->load($body)
+            && $validator->validate($loginForm)->isValid()
+        ) {
+            /** @var User $identity */
+            $identity = $identityRepository->findByLogin($loginForm->getAttributeValue('login'));
 
-                $user = new User($body['login'], $body['password']);
+            if ($identity !== null) {
+                $loginForm->getFormErrors()->addError('password', 'Unable to register user with such username.');
```
Comment (samdark) [inline src/Controller/SignupController.php:52]: Shouldn't it be added to "login" field?

File: views/auth/login.php [hunk 4]
```diff
@@ -2,18 +2,20 @@
...
  * @var \Yiisoft\Translator\TranslatorInterface $translator
  * @var \Yiisoft\Router\UrlGeneratorInterface $urlGenerator
  * @var string $csrf
+ * @var FormModelInterface $formModel
  */
 
-$this->setTitle($translator->translate('layout.login'));
-
-$error = $error ?? null;
+$this->setTitle(Html::encode($translator->translate('layout.login')));
```
Comment (samdark) [inline views/auth/login.php:18]: We should never escape input. Only output should be escaped.

So... https://github.com/yiisoft/demo/blob/master/views/layout/main.php#L46 there we should encode but not here.

File: views/auth/login.php [hunk 5]
```diff
@@ -22,23 +24,33 @@
...
-    ]) ?>
-  </div>
-  <button type="submit" class="btn btn-primary"><?= $translator->translate('layout.submit') ?></button>
-</form>
+<div class="container py-5 h-100">
+    <div class="row d-flex justify-content-center align-items-center h-100">
+        <div class="col-12 col-md-8 col-lg-6 col-xl-5">
+            <div class="card border border-dark shadow-2-strong rounded-3">
+                <div class="card-header bg-dark text-white">
+                    <h1 class="fw-normal h3 text-center"><?= $this->getTitle() ?></h1>
```
Comment (samdark) [inline views/auth/login.php:32]:
```suggestion
                    <h1 class="fw-normal h3 text-center"><?= Html::encode($this->getTitle()) ?></h1>
```

File: views/layout/main.php [hunk 6]
```diff
@@ -37,99 +37,101 @@
...
-    <?php $this->head() ?>
-</head>
-<body>
-<?php
-$this->beginBody();
+    <head>
+        <meta charset="utf-8">
+        <meta http-equiv="X-UA-Compatible" content="IE=edge">
+        <meta name="viewport" content="width=device-width, initial-scale=1">
+        <title>Yii Demo<?= $this->getTitle() ? ' - ' . $this->getTitle() : '' ?></title>
```
Comment (samdark) [inline views/layout/main.php:47]:
```suggestion
        <title>Yii Demo<?= $this->getTitle() ? ' - ' . Html::encode($this->getTitle()) : '' ?></title>
```

File: views/signup/signup.php [hunk 7]
```diff
@@ -2,42 +2,50 @@
...
  * @var \Yiisoft\View\WebView $this
  * @var \Yiisoft\Translator\TranslatorInterface $translator
  * @var \Yiisoft\Router\UrlGeneratorInterface $urlGenerator
  * @var string $csrf
+ * @var FormModelInterface $formModel
  */
 
-$this->setTitle($translator->translate('Signup'));
-$error = $error ?? null;
+$this->setTitle(Html::encode($translator->translate('Signup')));
```
Comment (samdark) [inline views/signup/signup.php:18]:
```suggestion
$this->setTitle($translator->translate('Signup'));
```

File: views/signup/signup.php [hunk 8]
```diff
@@ -2,42 +2,50 @@
...
-  <div class="alert alert-danger" role="alert">
-    <?= Html::encode($error) ?>
-  </div>
-<?php endif ?>
+<div class="container py-5 h-100">
+    <div class="row d-flex justify-content-center align-items-center h-100">
+        <div class="col-12 col-md-8 col-lg-6 col-xl-5">
+            <div class="card border border-dark shadow-2-strong rounded-3">
+                <div class="card-header bg-dark text-white">
+                    <h1 class="fw-normal h3 text-center"><?= $this->getTitle() ?></h1>
```
Comment (samdark) [inline views/signup/signup.php:26]:
```suggestion
                    <h1 class="fw-normal h3 text-center"><?= Html::encode($this->getTitle()) ?></h1>
```

File: src/Controller/SignupController.php [hunk 9]
```diff
@@ -25,31 +27,31 @@ public function __construct(ViewRenderer $viewRenderer)
...
-                if ($identity !== null) {
-                    throw new \InvalidArgumentException('Unable to register user with such username.');
-                }
+        if (
+            $request->getMethod() === Method::POST
+            && $loginForm->load($body)
+            && $validator->validate($loginForm)->isValid()
+        ) {
+            /** @var User $identity */
+            $identity = $identityRepository->findByLogin($loginForm->getAttributeValue('login'));
```
Comment (BoShurik) [inline src/Controller/SignupController.php:49]: Is it possible to move this check inside validator's rule?

File: views/auth/login.php (hunk 4 above)
Comment (terabytesoftw) [reply views/auth/login.php:18]: done.

File: views/auth/login.php (hunk 5 above)
Comment (terabytesoftw) [reply views/auth/login.php:32]: done.

File: views/layout/main.php (hunk 6 above)
Comment (terabytesoftw) [reply views/layout/main.php:47]: done.

File: views/signup/signup.php (hunk 7 above)
Comment (terabytesoftw) [reply views/signup/signup.php:18]: done.

File: views/signup/signup.php (hunk 8 above)
Comment (terabytesoftw) [reply views/signup/signup.php:26]: done.

**Comment by terabytesoftw** [comment]
closed in favor https://github.com/yiisoft/demo/pull/408

**Pull request closed without merging**